infrasim/dae.py

```diff
@@ -0,0 +1,628 @@
+                        if drv is not None:
+                            copy_drv = drv.get_port_cfg(link['atta_phy'])
+                            _scsi_id = exp['start_scsi_id'] + link['phy']
+                            drvs.append(copy_drv)
+                            copy_drv['scsi-id'] = _scsi_id
+                            link['atta_scsi_id'] = _scsi_id
+                            break
+                if link['atta_type'] == _Const.SES_DEVICE:
+                    _scsi_id = exp['start_scsi_id'] + link['phy']
+                    link['atta_scsi_id'] = _scsi_id
+        return drvs
+
+    def __process_connector(self):
+        """
+            iterate through all controllers in the configuration, setup links
+            under each port of controller.
+        """
+        index = 0
+        for controller in self.__cfg:
+            for _item in self.__dae_arrays:
+                _item.reset_for_traversal()
+            seses_dict = []
+            if 'connectors' in controller:
+                hba_links = []
+                file_name = "{0}/diskarray{1}.json".format(
+                    self.__ws_folder, index)
+                controller['dae_file'] = file_name
+                controller['drives'] = []
+                scsi_id_index = [0]
+                for link in controller['external_connectors']:
+                    # fill external link first.
+                    self.__fill_external_connector(link)
+                port_index = 0
+                for link in controller['connectors']:
+                    # fill connections to local hba port
+                    result = self.__traversal(
+                        seses_dict, port_index, link, scsi_id_index)
+                    for hba_link in hba_links:
+                        if _Const.check_range_conflict(result['phy'], result['phy_number'],
+                                                       hba_link['phy'], hba_link['phy_number']):
+                            raise Exception("HBA port duplicate use for phy {} to phy {}".format(
+                                result['phy'], result['phy'] + result['phy_number'] - 1))
+                    link.update(result)
+                    hba_links.append(
+                        {"phy": result['phy'], "phy_number": result['phy_number']})
+                    controller['drives'].extend(self.__list_drv_nodes(result))
+                    port_index = port_index + 1
+                controller['seses'] = seses_dict
+            index = index + 1
+
+    def create_dae_node(self, args):
+        """
+            dump a controller's port connection information and drive list to
+            a json file, which will be parsed to qemu command line.
+            args[0]: drive option list under a controller
+            args[1]: controller's configuration in dictionary
+        """
+        if args is None:
+            return
+        drv_args = []
+        drv_opt_list = args[0]
+        controller_info = args[1]
```

Comment:
🛠 Prefer to use keyword args here instead of a tuple, which is better to read and more pythonic.

Reply:
will change it.